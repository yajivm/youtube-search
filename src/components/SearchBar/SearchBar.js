import React, { useState, useEffect, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';

import SearchIcon from '../../assets/img/search.svg';
import CrossIcon from '../../assets/img/cross.svg';

import './SearchBar.style.css';

const SearchBar = ({
    onSearch,
    searchInputRef,
    canClearSearchInputValue,
    setClearSearchInputValue,
}) => {
    const [searchInputValue, setSearchInputValue] = useState('');

    const onSearchInputValueChanged = event => {
        const value = event.target.value;
        setSearchInputValue(value);
    }

    const clearSearchInputValue = () => {
        setSearchInputValue('');
    };

    const handleSubmit = event => {
        event.preventDefault();
        if (searchInputValue?.length) {
            onSearch(searchInputValue);
        }
    };

    useImperativeHandle(searchInputRef, () => ({
        getSearchKey: () => {
          return searchInputValue;
        }
    }));

    useEffect(() => {
        if (canClearSearchInputValue){
            setSearchInputValue('');
            setClearSearchInputValue(false);
        }
    }, [canClearSearchInputValue, setClearSearchInputValue]);

    return(
        <>
            <div className="search-input-wrapper" data-testid="search-input-wrapper">
                <form onSubmit={handleSubmit} className="search-form" data-testid="search-form">
                    <div className="form-controls">
                        <label>Search</label>
                        <input
                            id="video-search"
                            data-testid="video-search"
                            type="text"
                            value={searchInputValue}
                            onChange={onSearchInputValueChanged}
                            placeholder="Search"
                        />
                        {searchInputValue?.length ?
                            <div className="cross-icon-wrapper" data-testid="cross-icon" onClick={clearSearchInputValue}>
                                <img src={CrossIcon} alt="Icon" title="Clear" className="cross-icon" />
                            </div>
                        :
                            <></>
                        }
                    </div>
                </form>
                <button className="search-button" data-testid="search-button" onClick={handleSubmit}>
                    <img src={SearchIcon} alt="searchVideos" title="search" className="search-icon" />
                </button>
            </div>
        </>
    );
}

SearchBar.propTypes = {
    onSearch: PropTypes.func.isRequired,
    searchInputRef: PropTypes.shape({}),
    canClearSearchInputValue: PropTypes.bool,
};

SearchBar.defaultProps = {
    searchInputRef: null,
    canClearSearchInputValue: false,
};

export default SearchBar;
