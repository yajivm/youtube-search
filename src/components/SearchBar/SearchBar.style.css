.search-input-wrapper {
    height: 40px;
    display: flex;
    align-items: center;
}

.search-form {
    display: flex;
    justify-content: flex-start;
    padding: 0 20px;
    align-items: center;
    width: 100%;
}

.form-controls {
    display: flex;
    flex-grow: 1;
    align-items: center;
    justify-content: space-between;
}

.form-controls input[type="text"] {
    display: flex;
    border: 1px solid #3a363b;
    background-color: #121212;
    color: #fff;
    flex-grow: 1;
    padding: 10px;
    border-radius: 1px;
    font-size: 16px;
}

.form-controls input[type="text"]:focus {
    outline: none;
    border-color: #3a363b;
}

.form-controls label {
    display: none;
}

.search-button {
    height: 40px;
    background-color: #3a363b;
    border: 1px solid #3A363B;
    position: absolute ;
    right: 0;
    top: 0;
    cursor: pointer;
}

.search-icon {
    height: 30px;
}

.cross-icon-wrapper {
    position: absolute;
    right: 55px;
    cursor: pointer;
}

.cross-icon {
    width: 15px;
    height: 20px;
}
