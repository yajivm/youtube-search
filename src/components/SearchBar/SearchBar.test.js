import React, { createRef } from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import SearchBar from './SearchBar';

const mockOnSearch = jest.fn();
const mockRef = createRef();

test('should render SearchBar component', () => {
    const { getByTestId } = render(<SearchBar onSearch={mockOnSearch} searchInputRef={mockRef} />);
    
    const searchBarElement = getByTestId('search-input-wrapper');

    expect(searchBarElement).toBeDefined();
    expect(searchBarElement).toMatchSnapshot();
});

test('should mobile search bar when click enter keyboard', () => {
    const { getByTestId } = render(<SearchBar onSearch={mockOnSearch} searchInputRef={mockRef} />);
    
    const searchInput = getByTestId('video-search');
    
    expect(searchInput).toBeDefined();

    act(() => {
        fireEvent.change(searchInput, { target: { value: 'aTestWord' }});
    });

    expect(searchInput.value).toStrictEqual('aTestWord');

    fireEvent.submit(getByTestId('search-form'));

    expect(mockOnSearch).toHaveBeenCalledTimes(1);
    expect(mockOnSearch).toHaveBeenCalledWith('aTestWord');
});

test('should clear input value when click cross icon', () => {
    const { getByTestId } = render(<SearchBar onSearch={mockOnSearch} searchInputRef={mockRef} />);
    
    const searchInput = getByTestId('video-search');
    
    expect(searchInput).toBeDefined();

    act(() => {
        fireEvent.change(searchInput, { target: { value: 'aTestWord' }});
    });

    expect(searchInput.value).toStrictEqual('aTestWord');

    fireEvent.click(getByTestId('cross-icon'));

    expect(searchInput.value).toStrictEqual('');
});

test('should mobile search bar when enter data in input and click search icon', () => {
    jest.resetAllMocks();
    const { getByTestId } = render(<SearchBar onSearch={mockOnSearch} searchInputRef={mockRef} />);
    
    const searchInput = getByTestId('video-search');
    
    expect(searchInput).toBeDefined();

    act(() => {
        fireEvent.change(searchInput, { target: { value: 'aTestWord' }});
    });

    expect(searchInput.value).toStrictEqual('aTestWord');

    act(() => {
        fireEvent.click(getByTestId('search-button'));
    });

    expect(mockOnSearch).toHaveBeenCalledTimes(1);
    expect(mockOnSearch).toHaveBeenCalledWith('aTestWord');
});