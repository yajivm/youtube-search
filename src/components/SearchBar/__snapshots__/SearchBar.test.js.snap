// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchBar component 1`] = `
<div
  class="search-input-wrapper"
  data-testid="search-input-wrapper"
>
  <form
    class="search-form"
    data-testid="search-form"
  >
    <div
      class="form-controls"
    >
      <label>
        Search
      </label>
      <input
        data-testid="video-search"
        id="video-search"
        placeholder="Search"
        type="text"
        value=""
      />
    </div>
  </form>
  <button
    class="search-button"
    data-testid="search-button"
  >
    <img
      alt="searchVideos"
      class="search-icon"
      src="search.svg"
      title="search"
    />
  </button>
</div>
`;
